tests/benchmarks/test_discriminated_unions.py

```diff
@@ -0,0 +1,56 @@
+from __future__ import annotations
+
+from typing import Literal, Union
+
+from typing_extensions import Annotated
+
+from pydantic import BaseModel, Field, TypeAdapter
+
+
+def test_schema_build() -> None:
+    class NestedState(BaseModel):
+        state_type: Literal['nested']
+        substate: AnyState
+
+    class LoopState(BaseModel):
+        state_type: Literal['loop']
+        substate: AnyState
+
+    class LeafState(BaseModel):
+        state_type: Literal['leaf']
+
+    AnyState = Annotated[Union[NestedState, LoopState, LeafState], Field(..., discriminator='state_type')]
+    adapter = TypeAdapter(AnyState)
+    assert adapter.core_schema['schema']['type'] == 'tagged-union'
+
+
+def build_nested_state(n):
+    if n <= 0:
+        return {'state_type': 'leaf'}
+    else:
+        return {'state_type': 'loop', 'substate': {'state_type': 'nested', 'substate': build_nested_state(n - 1)}}
+
+
+class NestedState(BaseModel):
+    state_type: Literal['nested']
+    substate: AnyState
+
+
+class LoopState(BaseModel):
+    state_type: Literal['loop']
+    substate: AnyState
+
+
+class LeafState(BaseModel):
+    state_type: Literal['leaf']
+
+
+AnyState = Annotated[Union[NestedState, LoopState, LeafState], Field(..., discriminator='state_type')]
+adapter = TypeAdapter(AnyState)
+
+
+def test_efficiency_with_highly_nested_examples() -> None:
+    # can go much higher, but we keep it reasonably low here for a proof of concept
+    for i in range(1, 12):
+        very_nested_input = build_nested_state(i)
+        adapter.validate_python(very_nested_input)
```

Comment:
Can we separate the build and runtime tests so that one doesn't test the other?

Reply:
The build_nested_state here just creates mock data, so not sure if we need separation... lmk if I'm misunderstanding!

Reply:
Ah okay disregard then I misunderstood